libcore/src/geometry/helper/CorrectGeometry.cpp

```diff
@@ -0,0 +1,354 @@
+        }
+    }
+    return false;
+}
+
+/**
+ * @brief Checks if the wall is inside the polygon of the subroom and does not exist already
+ */
+bool IsConnectedAndNewWall(const SubRoom & subroom, const Wall & wall)
+{
+    if(subroom.GetAllWalls().end() !=
+       std::find(subroom.GetAllWalls().begin(), subroom.GetAllWalls().end(), wall)) {
+        return false; // The wall already exists in subroom (not new).
+    }
+
+    if(isPointAndSubroomIncident(subroom, wall.GetPoint1()) &&
+       isPointAndSubroomIncident(subroom, wall.GetPoint2())) {
+        return true;
+    }
+
+
+    return false;
+}
+
+/**
+ * @brief Adds wall pieces to a subroom if they are new and connected
+ */
+int AddWallToSubroom(SubRoom & subroom, const std::vector<Wall> & wallPieces)
+{
+    int wallsAdded = 0;
+
+    for(const auto & w : wallPieces) {
+        if(IsConnectedAndNewWall(subroom, w)) {
+            subroom.AddWall(w);
+            wallsAdded++;
+        }
+    }
+
+    return wallsAdded;
+}
+
+/**
+ * @brief Replace BigWall with some of the wallPieces
+ */
+void ReplaceBigWall(SubRoom & subroom, const Wall & bigWall, const std::vector<Wall> & wallPieces)
+{
+    Logging::Debug(fmt::format(
+        check_fmt("Replacing big wall {} in SubRoom {}-{}"),
+        bigWall.toString(),
+        subroom.GetRoomID(),
+        subroom.GetSubRoomID()));
+
+    if(!subroom.RemoveWall(bigWall)) {
+        throw std::runtime_error(fmt::format(
+            check_fmt("Correcting the geometry failed. Could not remove wall: {}"),
+            bigWall.toString()));
+    }
+
+    if(AddWallToSubroom(subroom, wallPieces) <= 0) {
+        throw std::runtime_error(fmt::format(
+            check_fmt("Correcting the geometry failed. Could not add any wall from splitted big "
+                      "wall: {}"),
+            bigWall.toString()));
+    }
+}
+
+bool RemoveBigWalls(SubRoom & subroom)
+{
+    bool wallsRemoved = false;
+
+    // Need to copy walls here, because we are going to change subrooms wall container
+    auto walls = subroom.GetAllWalls();
+
+    for(auto const & bigWall : walls) {
+        auto splitWallPieces = SplitWall(subroom, bigWall);
+
+        if(splitWallPieces && !splitWallPieces->empty()) {
+            // remove big wall and add one wallpiece to walls
+            ReplaceBigWall(subroom, bigWall, *splitWallPieces);
+            wallsRemoved = true;
+        }
+    }
+
+    return wallsRemoved;
+}
+
+
+} // namespace
+
+void correct(Building & building)
+{
+    auto t_start = std::chrono::high_resolution_clock::now();
+    Logging::Info("Starting geometry::helper::correct to fix the geometry ...");
+
+    bool geometry_changed = false;
+
+    for(const auto & room : building.GetAllRooms()) {
+        for(const auto & keySubroomPair : room.second->GetAllSubRooms()) {
+            SubRoom & subroom = *keySubroomPair.second;
+
+            // remove wall parts overlapping with exits
+            geometry_changed = RemoveOverlappingDoors(subroom);
```

Comment:
This looks wrong, `geometry_changed` is overwritten.
Needs tests :)

Reply:
Done.